itou/job_applications/notifications.py

```diff
@@ -1,17 +1,83 @@
+    def __init__(self, job_application=None):
+        active_memberships = None
+        if job_application:
+            active_memberships = job_application.to_siae.siaemembership_set.active()
+        super().__init__(recipients_qs=active_memberships)
+        self.job_application = job_application
+
+    @property
+    def name(self):
+        return "new_spontaneous_job_application_siae_email"
+
+    @property
+    def email(self):
+        to = self.recipients_emails
+        context = {"job_application": self.job_application}
+        subject = "apply/email/new_for_siae_subject.txt"
+        body = "apply/email/new_for_siae_body.txt"
+        return get_email_message(to, context, subject, body)
+
+    @property
+    def recipients_emails(self):
+        return self.get_recipients().values_list("user__email", flat=True)
+
+
+class NewQualifiedJobAppSiaeNotification(NotificationBase):
     """
-    Notification sent to employers after a job seeker applied.
+    Subscribe a recipient to a job description and send notifications.
+    A job description represents an SiaeJobDescription object also known as
+      a `job_application.selected_jobs` relation.
+
+    ----
+
+    Usage:
+    ## Send a notification from a job_application which have selected_jobs
+    ```
+    notification = cls(job_application=job_application)
+    notification.send()
+    ```
+
+    ## Opt in or opt out a recipient to a list of job descriptions
+    ```
+    selected_jobs_pks = JobDescription.objects.filter().values_list("pk", flat=True)
+    notification = cls(selected_jobs_pks=selected_jobs_pks)
+    notification.subscribe(recipient)
+
+    # unsubscribe the same way:
+    notification.unsubscribe(recipient)
+    ```
+
+    ## Know if a recipient opted out to a specific job description notifications
+    ```
+    selected_jobs_pks = [JobDescription.objects.get()]
+    notification = cls(selected_jobs_pks=selected_jobs_pks)
+    notification.is_subscribed(recipient)
+    ```
     """
 
-    def __init__(self, job_application):
-        active_memberships = job_application.to_siae.siaemembership_set.active()
-        super().__init__(recipients_qs=active_memberships)
+    def __init__(self, selected_jobs_pks=None, job_application=None):
+        """
+        Arguments:
+        - selected_jobs_pk: list of integers. It represents an SiaeJobDescription object
+        (also known as a job_application.selected_jobs relation). If empty, job_application is mandatory.
+        - job_application: a JobApplication object. If empty, selected_jobs_pks is mandatory.
+        """
+        active_memberships = None
         self.job_application = job_application
-        self.siae = job_application.to_siae
+        self.subscribed_pks = selected_jobs_pks
+        if job_application:
+            active_memberships = job_application.to_siae.siaemembership_set.active()
+            self.subscribed_pks = job_application.selected_jobs.values_list("pk", flat=True)
```

Comment:
J'ai l'impression qu'il y a un problème d'interface ici, ça rend la lecture un peu bizarre car on mélange `self.subscribed_pks` avec 2 concepts :

- une fois pour savoir à quelles fiches de postes on doit `opt-in` ou `opt-out` un utilisateur
- une fois pour savoir quelles fiches de postes sont concernées par la notification

Or, on est bien dans une notification qui devrait se contenter d'avoir un paramètre `job_application` et lever cette ambiguïté.

J'ai une idée pour découpler notification et *opt-in/out* sans en faire deux entités, mais je ne sais pas si ça va te convenir : peut-être qu'on devrait utiliser des [`@staticmethod`s](https://docs.python.org/3/library/functions.html#staticmethod) pour simplement attacher certaines fonctions comme `subscribe`, `unsubscribe` etc. à la classe ?

Ça nécessite de passer `NAME` et `SUB_NAME` en variables de classe plutôt qu'en _properties_ :

```python
@staticmethod
def subscribe(recipient, selected_jobs_pks):

    pks_set = NewQualifiedJobAppSiaeNotification._get_recipient_subscribed_pks(recipient)
    pks_set.update(selected_jobs_pks)

    name = NewQualifiedJobAppSiaeNotification.NAME
    sub_name = NewQualifiedJobAppSiaeNotification.SUB_NAME

    recipient.notifications[name][sub_name] = list(pks_set)
    recipient.save()
```

Et en sortie de l'UI où tu enregistres la configuration de l'utilisateur :

```python
NewQualifiedJobAppSiaeNotification.subscribe(recipient, selected_jobs_pks)
```

Reply:
SUPER IDÉE ! Je teste cela tout de suite.

Reply:
Alors... Le souci est qu'il devient difficile de surcharger les méthodes parent. 

```
class NotificationBase:
    NAME = None

    @staticmethod
    def is_subscribed(recipient):
        name = NotificationBase.NAME
        return recipient.notifications.get(name) and recipient.notifications[name]["subscribed"]

class NewQualifiedJobAppSiaeNotification(NotificationBase):
      NAME = "new_qualified_job_application_siae_email"
```

```
NewQualifiedJobAppSiaeNotification.is_subscribed(recipient) 
# => returns None as NAME is NotificationBase.NAME and not cls.NAME
```

Le meilleur moyen est d'utiliser une méthode de classe, mais ça ne va pas te plaire...

Reply:
L'idée ça n'est pas que ça me plaise mais qu'on comprenne tout de suite le code en le lisant 😄 

Peut-être qu'il faut continuer à passer par des méthodes de classe mais clarifier les noms des variables pour que les responsabilités et l'interface soit claire.